docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Variable prefixes
+
+| Prefix |Description |
+|--------|------------|
+| `ce`   | Official CloudEvents attributes defined in the spec, plus extension attributes if requested. |
+| `data` | Fields parsed from the event data, if requested. |
+
+Official attributes and extension attributes are both available to the
+expression under the prefix `ce`. This co-mingling is intended to ease the
+promotion of extension attributes to official attributes. The prefix `ce` was
+chosen for succintness and its suggestion that CloudEvents is the operative
+specification.
+
+Data fields are available to the expression under the prefix `data`.
+
+Parsing dynamic extension attributes and data fields may be costly. In this
+initial proposal, the Trigger must explicitly enable dynamic parsing by setting
+boolean fields `parseExtensions` and/or `parseData`. Future work may
+eliminate the need for these fields by detecting the use of extension attributes
+and/or data fields in the expression.
+
+### Why CEL?
+
+CEL is the
+[Common Expression Language](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/doc/langdef.md).
+It is designed for expressing security policy and protocols, and has the
+following
+[guiding philosophy](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/README.md):
+
+*   **Small and fast.** CEL is not Turing complete, so it is easy to understand
+    and implement.
+*   **Extensible.** CEL is designed to be embedded and can be extended with new
+    functionality.
+*   **Developer-friendly.** The language spec was based on experience and
+    usability testing from implementing Firebase Rules.
+
+Additional properties of CEL that make it attractive for Trigger Filtering
+include:
+
+*   **Familiar expression syntax.** Filter expressions written in CEL look and
+    behave like conditional expressions written in the most popular languages.
+*   **Expressions can be compiled and cached for efficiency.** We expect users
+    will create many filters that change rarely in comparison to frequency of
+    evaluations. This usage pattern works in favor of an expression language
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
```

Comment:
Should this include a version key? How will updates and deprecations to the trigger filtering api be handled?

Reply:
I think we can rely on CRD versioning to version the yaml part of the API (the fields allowed in the `spec` of the Trigger object).

I would hope that we could also rely on CRD versioning to identify the version of expression API. I think it would be easier for both users and implementers if there is a single version string to identify the behavior of the entire Trigger.

For the expression API, we _could_ add a version field adjacent to the `expression` field in case the expression semantics change without a change to the yaml spec. I haven't thought about this too carefully, so I may be wrong, but I don't like how this complicates Trigger versioning. Because there are two version fields involved, it seems like users and implementers must understand the combination of those two versions to understand how each Trigger will behave.

```yaml
# Notice this is v1beta1, the version that follows v1alpha1
apiVersion: eventing.knative.dev/v1beta1
kind: Trigger
spec:
  filter:
    cel:
      expression: ce.type == "com.example.event"
      # What happens when v2 is specified for a v1alpha1 Trigger?
      version: v2
```

An argument could be made that `apiVersion` is the version of the _format_ and `spec.filter.cel.version` is the version of the _behavior_. I could buy this, but I'd want to see some examples from other Kubernetes objects.

It may eventually become necessary to add fields that specify expression behavior, e.g. expression error handling. These could be adjacent to the expression field:

```yaml
spec:
  filter:
    cel:
      expression: ce.type == "com.example.event"
      onError: accept # alternatively "reject"
```

I'm more ok with these than `spec.filter.cel.version` since they are more explicitly about behavior, and they hint at what behavior they specify, but I'm not sold on them for the same reason as above: they would be tempting to add, and having a lot of them could make Triggers hard to interpret and implement.

Reply:
Thanks for the thorough response. I think it's fair to tie a) the specific CEL version and b) the format of the trigger yaml to the `apiVersion` field for the entire trigger. It would be more complex to manage triggers where the version of CEL is not specifically tied to a version of the trigger. It does mean users cannot bring their own CEL version independent of the yaml structure when implementing an expression, but on the other hand users get a more understandable development experience, and more maintainable from the knative side. I think this is a reasonable tradeoff. The main concern I could see is a change in CEL could block users from upgrading because they need to upgrade their expression at the same time as knative, which is a more risky upgrade. The end result might be that knative is explicit in which version of CEL is used in each version.

Reply:
Good point about the upgrade concern. I added a caveat to the document discussing the CEL versioning strategy.

I think the worst case is:

- Multiple incompatible CEL versions
- Expressions can't be upgraded automatically
- The CEL version of an expression can't be identified

I think in this case we must add a version specifier field, at least for one version of Knative, and run both versions of the CEL runtime concurrently.